### PR TITLE
Add annotation item resizing keeping aspect ratio

### PR DESCRIPTION
**Pull request by fnkabit**
Issue https://github.com/ksnip/ksnip/issues/484

**Work in progress**

### CODE REVIEW

**Comment by fnkabit** [comment]
@DamirPorobic I think it's usable at this point, do you think I should do some refactoring first ?

File: src/annotations/items/AbstractAnnotationItem.h [hunk 1]
```diff
@@ -42,7 +42,7 @@ class AbstractAnnotationItem : public QGraphicsWidget
 	QPainterPath shape() const override;
 	virtual bool intersects(const QRectF &rect) const;
 	virtual void addPoint(const QPointF &position, bool modified) = 0;
-	virtual void setPointAt(const QPointF &point, int index) = 0;
+    virtual void setPointAt(const QPointF &point, int index, bool keepAspectRatio = false) = 0;
```
Comment (DamirPorobic) [inline src/annotations/items/AbstractAnnotationItem.h:45]: I'm personally not longer a fan of default values and I think it's not recommend to use them in virtual methods. I would suggest to either always specify the keepAspectRatio value or have two method with different parameters.

File: src/annotations/items/AbstractAnnotationLine.cpp [hunk 2]
```diff
@@ -57,7 +57,7 @@ QLineF AbstractAnnotationLine::line() const
 	return *mLine;
 }
 
-void AbstractAnnotationLine::setPointAt(const QPointF &point, int index)
+void AbstractAnnotationLine::setPointAt(const QPointF &point, int index, bool keepAspectRatio)
```
Comment (DamirPorobic) [inline src/annotations/items/AbstractAnnotationLine.cpp:60]: For line it's not really aspectRatio but snapping to angles, maybe we could implement that here too?

File: src/common/helper/ShapeHelper.cpp [hunk 3]
```diff
@@ -18,6 +18,8 @@
  */
 
 #include <QtGui/QPainterPath>
+#include <QtMath>
+#include <QtDebug>
```
Comment (DamirPorobic) [inline src/common/helper/ShapeHelper.cpp]: For what is QtDebug used?

File: src/common/helper/ShapeHelper.cpp [hunk 4]
```diff
@@ -127,23 +129,59 @@ QPointF ShapeHelper::rectPointAtIndex(const QRectF &rect, int index)
...
-		updatedRect.setTopLeft(pos);
+		if (!keepAspectRatio) {
+			updatedRect.setTopLeft(pos);
+		} else {
+			const auto xDif = updatedRect.topLeft().x() - pos.x();
+			const auto yDif = updatedRect.topLeft().y() - pos.y();
+			const auto min = qMin(xDif, yDif);
+			const auto newX = updatedRect.topLeft().x() - min;
+			const auto newY = updatedRect.topLeft().y() - min;
+			updatedRect.setTopLeft(QPointF(newX, newY));
```
Comment (DamirPorobic) [inline src/common/helper/ShapeHelper.cpp:143]: Don't we have some similar logic in the addPoint for Rect? I believe we can already keep aspect ratio when drawing the initial rect, could that have somehow been used for this feature? If not, could extract this logic in a separate class? I imagine we could have something like a RectShapeUpdater that gets just a rect and a bool for the aspectRatio as parameter and either updates that rect or returns a new updated rect  (probably rather the later). You would have here only a call to this class and this class could be simply covered by unit tests.

**Review by DamirPorobic — COMMENTED** [review]
Looking good, I would just add the snap to angle for line as we have it for drawing initial line just like we have keep aspect ratio when drawing the initial rect. For the initial stuff I have called the boolean "modifier" but it doesn't really explain what it's doing, maybe you you have some better idea for a term that covers the snap to line and keep aspect ratio

File: src/common/helper/ShapeHelper.cpp (hunk 3 above)
Comment (fnkabit) [reply src/common/helper/ShapeHelper.cpp]: Has been removed in the meantime

**Pull request merged**